### PR TITLE
Closes #2412 - tasks jobs log improvements

### LINKED ISSUE #2412: Plugin main page is very long to display
Currently, on my test server the plugin main page need **almost 30 seconds** to get displayed.

I found out that this is because of a very huge and not optimized SQL query built by the `Task::getJobLogs` function.

The SQL request:
```
SELECT task.`id` AS 'task.id', task.`name` AS 'task.name', job.`id` AS 'job.id', job.`name` AS 'job.name', job.`method` AS 'job.method', agent.`id` AS 'agent.id', agent.`name` AS 'agent.name', agent.`computers_id` AS 'agent.computers_id', run.`id` AS 'run.id', run.`itemtype` AS 'run.itemtype', run.`items_id` AS 'run.items_id', run.`state` AS 'run.state', log.`date` AS 'log.last_date', UNIX_TIMESTAMP(log.`date`) AS 'log.last_timestamp', log.`id` AS 'log.last_id', log.`comment` AS 'log.last_comment'
FROM `glpi_plugin_fusioninventory_agents` AS agent
INNER JOIN ( SELECT MAX(run.`id`) AS max_id, run.`plugin_fusioninventory_agents_id`, run.`plugin_fusioninventory_taskjobs_id`, run.`items_id`, run.`itemtype`, MAX(log.`id`) AS max_log_id
FROM ( SELECT *
FROM ( SELECT *, @num := IF(@agent_id = plugin_fusioninventory_agents_id && @taskjob_id = plugin_fusioninventory_taskjobs_id && @items_id = items_id && @itemtype = itemtype, @num:= @num + 1, 1) AS row_num, @agent_id:=plugin_fusioninventory_agents_id AS tmp_var1, @taskjob_id:=plugin_fusioninventory_taskjobs_id AS tmp_var2, @items_id:=items_id AS tmp_var3, @itemtype:=itemtype AS tmp_var4
FROM glpi_plugin_fusioninventory_taskjobstates
ORDER BY plugin_fusioninventory_taskjobs_id, plugin_fusioninventory_agents_id, itemtype, items_id, id DESC ) AS limited_states
WHERE row_num <= 2 ) AS run
LEFT JOIN `glpi_plugin_fusioninventory_taskjoblogs` AS log ON log.`plugin_fusioninventory_taskjobstates_id` = run.`id` GROUP BY run.`plugin_fusioninventory_agents_id`, run.`plugin_fusioninventory_taskjobs_id`, run.`items_id`, run.`itemtype`, run.`id` ) max_run ON max_run.`plugin_fusioninventory_agents_id` = agent.`id`
INNER JOIN `glpi_plugin_fusioninventory_taskjobstates` AS run ON max_run.`max_id` = run.`id`
LEFT JOIN `glpi_plugin_fusioninventory_taskjoblogs` as log ON max_run.`max_log_id` = log.`id`
INNER JOIN `glpi_plugin_fusioninventory_taskjobs` AS job ON job.`id` = run.`plugin_fusioninventory_taskjobs_id`
INNER JOIN `glpi_plugin_fusioninventory_tasks` as task ON job.`plugin_fusioninventory_tasks_id` = task.`id`
WHERE 1 AND task.`id` IN (0,6,12,13,14,15) GROUP BY job.`id`, agent.`id`, run.`id`, log.`id`
ORDER BY agent.`id` ASC, run.`id` DESC;
```

The EXPLAIN of this request:
image

I am not enough SQL aware to propose an optimization of this request 😞 Anyone ?

### PR DESCRIPTION
**Pull request by mohierf**
Improve the huge query used to get the tasks jobs log (closes #2412)
Add some extra-debug logs for this process
Allow to get information withou or without execution result log
Allow to view graph/log for non active tasks (closes #2314)

### CODE REVIEW

**Comment by mohierf** [comment]
**Note** that currently the jobs log contain only the last job execution. As such, the *Include old jobs* is no more active!

**Comment by mohierf** [comment]
@ddurieux : waiting for your improvements on the SQL requests

File: inc/task.class.php [hunk 1]
```diff
@@ -923,80 +936,98 @@ function forceRunning() {
...
-         ['job.id'      , 'job.`id`'],
-         ['job.name'    , 'job.`name`'],
-         ['job.method'  , 'job.`method`'],
-         ['job.targets' , 'job.`targets`'],
-      ];
-
-      $fieldmap = [];
-      foreach ($query_fields as $index => $key) {
-         $fieldmap[$key[0]]=$index;
+      $entity_restrict_task = '';
```
Comment (wawax) [inline inc/task.class.php]: Please add a comment to explain why you're restricting the query

File: inc/task.class.php [hunk 2]
```diff
@@ -923,80 +936,98 @@ function forceRunning() {
...
+      // Target cache (used to speed up data formatting)
       $expanded      = [];
       if (isset($_SESSION['plugin_fusioninventory_tasks_expanded'])) {
          $expanded = $_SESSION['plugin_fusioninventory_tasks_expanded'];
       }
 
-      while ($result = $data_structure['result']->fetch_row()) {
-         $task_id = $result[$fieldmap['task.id']];
+      $agent_state_types = [
+         'agents_prepared', 'agents_running', 'agents_cancelled',
```
Comment (wawax) [inline inc/task.class.php]: Agent states are used several times in the code. Please consider adding class constants

File: inc/task.class.php [hunk 3]
```diff
@@ -1071,311 +1102,280 @@ function getJoblogs($task_ids = []) {
...
+               PluginFusioninventoryToolbox::logIfExtradebug(
+                  "pluginFusioninventory-tasks",
+                  "Log query: " . print_r($q_job_state_last_log, true));
+               while ($log_result = $q_job_state_last_log['result']->fetch_assoc()) {
+                  PluginFusioninventoryToolbox::logIfExtradebug(
+                     "pluginFusioninventory-tasks",
+                     "Log: " . print_r($log_result, true));
+                  $targets[$target_id]['agents'][$agent_id][] = [
+                     'agent_id'      => $agent_id,
+                     'link'          => $CFG_GLPI['root_doc']."/front/computer.form.php?id=" .$result['agent.computers_id'],
```
Comment (wawax) [inline inc/task.class.php]: please use Computer::getFormURLWithID

File: inc/task.class.php [hunk 4]
```diff
@@ -1071,311 +1102,280 @@ function getJoblogs($task_ids = []) {
...
-                  'numstate'      => $result[$fieldmap['run.state']],
+                  'link'          => $CFG_GLPI['root_doc']."/front/computer.form.php?id=" .$result['agent.computers_id'],
+                  'numstate'      => $result['state'],
                   'state'         => $agent_state,
                   'jobstate_id'   => $run_id,
-                  'last_log_id'   => $result[$fieldmap['log.last_id']],
-                  'last_log_date' => $result[$fieldmap['log.last_date']],
-                  'timestamp'     => $result[$fieldmap['log.last_timestamp']],
-                  'last_log'      => $result[$fieldmap['log.last_comment']]
+                  //'last_log_id'   => $result['log.last_id'],
```
Comment (wawax) [inline inc/task.class.php]: Please remove this commented code

File: inc/task.class.php [hunk 5]
```diff
@@ -1464,11 +1478,11 @@ static function getItemsFromDB($filter) {
       if (isset( $filter['is_running'])
               AND is_bool($filter['is_running'])) {
          //TODO: get running taskjobs
-//         if ($filter['is_running']) {
-//            $where[] = "( task.`execution_id` != taskjob.`execution_id` )";
-//         } else {
-//            $where[] = "( task.`execution_id` = taskjob.`execution_id` )";
-//         }
+         //if ($filter['is_running']) {
```
Comment (wawax) [inline inc/task.class.php]: Please remove this commented code

File: inc/task.class.php [hunk 6]
```diff
@@ -1548,13 +1562,13 @@ static function getItemsFromDB($filter) {
       }
 
       //TODO: Filter by list of IDs
-      if (isset($filter['by_ids'])
-              AND is_bool($filter['by_entities'])) {
-      }
+      //if (isset($filter['by_ids'])
```
Comment (wawax) [inline inc/task.class.php]: Please remove this commented code

File: inc/task.class.php [hunk 7]
```diff
@@ -750,11 +757,14 @@ public function getAgentsFromActors($actors = [], $use_cache = false) {
              * TODO: The following should be replaced with Dynamic groups
              */
             case 'PluginFusioninventoryAgent':
-               switch($itemid) {
+               switch ($itemid) {
+
                   case "dynamic":
```
Comment (wawax) [inline inc/task.class.php]: I supposed that "dynamic" and "dynamic-same-subnet" should be safely removed right ?
In this case you can also remove the switch/case

File: inc/task.class.php [hunk 8]
```diff
@@ -1071,311 +1102,280 @@ function getJoblogs($task_ids = []) {
...
+                  // if we don't have success run (more recent due to previous test)
+                  // so we are really in error
+                  if (!isset($counters['agents_success'][$agent_id])) {
+                     $counters['agents_error'][$agent_id] = $run_id;
+                     unset($counters['agents_notdone'][$agent_id]);
+                  }
 
-                     $agent_state = 'error';
-                     break;
+                  $agent_state = 'error';
```
Comment (wawax) [inline inc/task.class.php]: Should this state be also a constant ?

File: inc/task.class.php [hunk 9]
```diff
@@ -1690,7 +1705,7 @@ function csvExport($params = []) {
       ];
       $params = array_merge($default_params, $params);
 
-      $includeoldjobs    = $_SESSION['glpi_plugin_fusioninventory']['includeoldjobs'];
+      $includeoldjobs = $_SESSION['glpi_plugin_fusioninventory']['includeoldjobs'];
       $agent_state_types = ['prepared', 'cancelled', 'running', 'success', 'error' ];
```
Comment (wawax) [inline inc/task.class.php]: In line 1040 we have another variable called $agent_state_types with other states... for example 'agent_prepared' and here 'prepared'
Could'nt we just normalize that and use only one constant value ?

**Pull request closed without merging**